rust/kernel/thread.rs

```diff
@@ -0,0 +1,508 @@
+///
+/// let arc = Arc::try_new(AtomicUsize::new(0))?;
+///
+/// let t = thread_try_new!(
+///   cstr!("rust-thread"),
+///   |x: Arc<AtomicUsize>| -> KernelResult<()> {
+///     for _ in 0..10 {
+///         x.fetch_add(1, Ordering::Release);
+///         println!("x is {}", x.load(Ordering::Relaxed));
+///     }
+///     Ok(())
+///   },
+///   arc.clone()
+/// )?;
+///
+/// t.wake_up();
+///
+/// while arc.load(Ordering::Acquire) != 10 {
+///     schedule();
+/// }
+///
+/// println!("main thread: x is {}", arc.load(Ordering::Relaxed));
+/// ```
+///
+/// # Context
+///
+/// This macro might sleep due to the memory allocation and waiting for
+/// the completion in `kthread_create_on_node`. Therefore do not call this
+/// in atomic contexts (i.e. preemption-off contexts).
+#[macro_export]
+macro_rules! thread_try_new {
+    ($name:expr, $func:expr, $arg:expr) => {{
+        // In case of failure, we need to `transmute` the `$arg` back, `_arg` is
+        // used here to inference the type of `$arg`, so that the `transmute`
+        // in the failure path knows the type.
+        let mut _arg = $arg;
+
+        // TYPE CHECK: `$arg` should be the same as `*mut c_void`, and
+        // `transmute` only works if two types are of the same size.
+        //
+        // SAFETY: In the bridge funciton, the `$arg` is `transmute` back.
+        let data = unsafe { core::intrinsics::transmute(_arg) };
+
+        // SAFTEY: a) the bridge function is a valid function pointer, and b)
+        // the bridge function `transmute` back what we just `transmute`.
+        let result =
+            unsafe { $crate::thread::Thread::try_new_c_style($name, $crate::bridge!($func), data) };
+
+        if let Err(e) = result {
+            // Creation fails, we need to `transmute` back the `$arg` because
+            // there is no new thread to own it, we should let the current
+            // thread own it.
+            //
+            // SAFETY: We `transmute` back waht we just `transmute`, and since
+            // the new thread is not created, so no one touches `data`.
+            unsafe {
+                _arg = core::intrinsics::transmute(data);
+            }
+
+            // Uses an unused closure to check whether `$arg` is `Send`.
+            let _ = move || _arg;
+
+            Err(e)
+        } else {
+            result
+        }
+    }};
+}
+
+/// Helper trait for thread arguments handling.
+///
+/// Basically, we just handle object "sending" manually, for a [`Send`] object,
+/// [`ThreadArg::from_arg`] is used to convert it into a pointer that kernel's
+/// thread creation function will accept (e.g. `kthread_create_on_node`). And
+/// later in the thread function, [`ThreadArg::from_raw`] is used to convert it
+/// back.
+///
+/// # Safety
+///
+/// Both [`ThreadArg::from_arg`] and [`ThreadArg::from_raw`] are marked as
+/// `unsafe` because they are only allowed to use in pair.
+pub trait ThreadArg<A: Send> {
```

Comment:
This has many similarities with `PointerWrapper`, we should see if we can merge them to avoid the proliferation of such wrappers.

Reply:
Cool! I wasn't aware of that, And actually I wasn't aware that I could `impl<T> ... for Box<T>` outside where `Box` is defined. ;-) I think we can put the `PointerWrapper` into `kernel::type` and share between submodules, do you plan to do that? For do you want a PR from me?

Reply:
I think another PR later is best. We can open an issue on that.

Reply:
Question: can I drop the type parameter for `trait PointerWrapper`? Like:
``` rust
pub trait PointerWrapper {
    fn into_pointer(self) -> *const c_types::c_void;
    unsafe fn from_pointer(ptr: *const c_types::c_void) -> Self;
}

impl<T> PointerWrapper for Box<T> {
    fn into_pointer(self) -> *const c_types::c_void {
        Box::into_raw(self) as _
    }
    unsafe fn from_pointer(ptr: *const c_types::c_void) -> Self {
        Box::from_raw(ptr as _)
    }
}
```

Because IIUC we don't care about the type that the pointer is pointing, we only want to store the pointer value somewhere, so we can get it back. And since we use the same `PointerWrapper`, we are guaranteed to get the correct type once we `from_pointer`.

Reply:
> Question: can I drop the type parameter for `trait PointerWrapper`?

Yes, I think you're right in that it's safe to drop the type parameter.